dlib/cuda/cuda_dlib.h

```diff
@@ -412,6 +412,62 @@ namespace dlib
 
     // ----------------------------------------------------------------------------------------
 
+        class compute_loss_binary_log_per_pixel
+        {
+            /*!
+                The point of this class is to compute the loss computed by
+                loss_binary_log_per_pixel, but to do so with CUDA.
+            !*/
+        public:
+
+            compute_loss_binary_log_per_pixel(
+            )
+            {
+            }
+
+            template <
+                typename const_label_iterator
+                >
+            void operator() (
+                const_label_iterator truth,
+                const tensor& subnetwork_output,
+                tensor& gradient,
+                double& loss
+            ) const
+            {
+                const size_t bytes_per_plane = subnetwork_output.nr()*subnetwork_output.nc()*sizeof(float);
+                // Allocate a cuda buffer to store all the truth images and also one float
+                // for the scalar loss output.
+                buf = device_global_buffer(subnetwork_output.num_samples()*bytes_per_plane + sizeof(float));
+
+                cuda_data_void_ptr loss_buf = buf;
+                buf = buf+sizeof(float);
```

Comment:
I just realized there isn't any method to cast a `cuda_data_void_ptr` to a `cuda_data_ptr<float>`, which is really what should happen here :|

Reply:
I just pushed `static_pointer_cast()`.  So you can do this now:
```
cuda_data_ptr<float> loss_buf = static_pointer_cast<float>(buf, 1);
```
and `loss_buf.get() == buf.get()` and `loss_buf.size() == 1`.  Then you can use the nicely typed loss_buf rather than void pointers.

Reply:
This should look better now.